osu.Server.Spectator/Hubs/MultiplayerHub.cs

```diff
@@ -227,6 +232,40 @@ public async Task ChangeState(MultiplayerUserState newState)
             }
         }
 
+        public async Task ChangeBeatmapAvailability(BeatmapAvailability newBeatmapAvailability)
+        {
+            using (var userUsage = await GetOrCreateLocalUserState())
+            using (var roomUsage = await getLocalUserRoom(userUsage.Item))
+            {
+                var room = roomUsage.Item;
+
+                if (room == null)
+                    throw new InvalidStateException("Attempted to operate on a null room");
+
+                var user = room.Users.Find(u => u.UserID == CurrentContextUserId);
+
+                if (user == null)
+                    throw new InvalidStateException("Local user was not found in the expected room");
+
+                if (user.State > MultiplayerUserState.Ready && user.State < MultiplayerUserState.Results)
+                    throw new InvalidStateException($"Attempted to change beatmap availability to {newBeatmapAvailability} while in {user.State} state");
+
+                if (user.BeatmapAvailability.Equals(newBeatmapAvailability))
+                    return;
+
+                user.BeatmapAvailability = newBeatmapAvailability;
+
+                // force state to idle when beatmap availability changes.
+                if (user.State != MultiplayerUserState.Idle)
+                {
+                    await Groups.RemoveFromGroupAsync(Context.ConnectionId, GetGroupId(room.RoomID, true));
+                    await changeAndBroadcastUserState(room, user, MultiplayerUserState.Idle);
+                }
```

Comment:
this... does not actually set `user.State`. please add test coverage for this.

also i 100% guarantee doing this locally like here will break at some point if changing state to `Idle` does more things than it currently does. which is why i'm not sure that reporting downloaded state should automatically back out the user from idle state server-side. maybe this should be two operations client-side (availability changed + user transitioned to idle).

Reply:
Fully agree on this. The download state was split out to *avoid* the complexity of changing the user's overall state, yet this change is adding weird intertwining rules which make things even harder to parse than if it was part of the main state enum.

Keep it as a display only state and don't require anything special from it. The other users in the room can choose whether to display the download state or not based on their own logic (ie. if the user is `Ready` the download state may be hidden even if the user is reporting downloading or not having the beatmap. or maybe it can just be left to display).

Reply:
Written it intentionally with safety checks such that nothing weird should ever be happening like a user playing while they don't own beatmap, but now that you all agree for the download states to be display-only, let's go with that.

Reply:
I'll apply changes once the dependency PR is merged first, as this is becoming a bit of a mess for me without the game package bumped.

Reply:
If anything it should be the reverse directionality - if a user becomes `Ready` their beatmap download state should be assumed to be `Available` and should silently update. But I'm not even sure that is required; would opt to keep things as simple as possible to avoid the logical overhead for something which doesn't need it.